fix: Build a new workshop for debugging using throw/try/catch/finally in js (#67750)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-library-search/6a1c49231334592cc5a4c6b0.md

---
id: 6a1c49231334592cc5a4c6b0
title: Step 7
challengeType: 0
dashedName: step-7
---

# --description--

Great! However, currently, if your `findBook` function throws an error, it will completely crash the script because nothing is there to catch it. 

To handle these errors gracefully, you need to wrap the code that might fail inside a `try` block. 

Look at the `searchBtn` event listener at the bottom of the file. Wrap all the existing code inside that callback function in a `try` block.

*Note: Ignore the red error line below the curly braces, you'll fix that in the next step.*
# --hints--

You should add a `try` block inside your event listener callback.

```js

> assert.match(code, /try\s*\{/);

Reviewer: how do you know it's in the right place?

Author: a behavioural test isn't possible here since there's no catch block yet and any thrown error crashes the script...the second regex verifies placement by checking both statements are wrapped inside try{} before the closing });, so placement is covered

---
id: 6a1c49231334592cc5a4c6b0
title: Step 5
challengeType: 0
dashedName: step-5
---

# --description--

Great! However, currently, if your `findBook` function throws an error, it will completely crash the script because nothing is there to catch it. 

> To handle these errors gracefully, you need to wrap the code that might fail inside a `try` block. 

Reviewer 2: "However, currently," stacks two adverbials before the clause even starts. Both lines also carry trailing whitespace.

```suggestion
Great! However, if your `findBook` function throws an error, it will completely crash the script because nothing is there to catch it.

To handle these errors gracefully, you need to wrap the code that might fail inside a `try` block.
```

> Look at the `searchBtn` event listener at the bottom of the file. Wrap all the existing code inside that callback function in a `try` block.

Reviewer 2: "all the existing code inside that callback function in a `try` block" stacks two prepositions, which makes it read as though the `try` block goes somewhere else.

```suggestion
Look at the `searchBtn` event listener at the bottom of the file. Wrap the existing code in that callback function in a `try` block.
```

> *Note: Ignore the red error line below the curly braces, you'll fix that in the next step.*

Reviewer: other than that, there is a whole error in the console

<img width="757" height="243" alt="image" src="https://github.com/user-attachments/assets/884135a9-0489-41a0-8a51-992d64249670" />

Author: i did notice this...but this step asks the camper to add the try block only, without the catch block, so it throws a red line...i mentioned it in a note below...is there an alternative you could suggest if this is an issue?

> *Note: Ignore the red error line below the curly braces, you'll fix that in the next step.*

Reviewer 2: Comma splice, and `you'll` is the only contraction in the workshop (step 1 has "you will build", step 3 has "You will fix this", step 9 has "You have successfully").

```suggestion
*Note: Ignore the red error line below the curly braces. You will fix that in the next step.*
```

# --hints--

> You should add a `try` block inside your event listener callback.

Reviewer 2: `/try\s*\{/` matches `try {` anywhere in the file, so this hint cannot check the "inside your event listener callback" part that the text promises. The hint below already grades placement, so either narrow the text or drop this hint as redundant.

```suggestion
You should add a `try` block.
```

```js
assert.match(code, /try\s*\{/);
```

Your `try` block should wrap the `const result` declaration and the `statusDisplay.textContent` assignment.

```js

> assert.match(code.replace(/\s+/g, ''), /try\{constresult=findBook\(bookInput\.value\.trim\(\)\);statusDisplay\.textContent=result;?\}\}\);/);

Reviewer 2: Two things here.

The trailing `\}\}\);` anchor requires the callback to close immediately after the `try` block, so it rejects submissions the editable region allows. Checked in Node against the whitespace-stripped code:

| Camper writes | Current | Proposed |
| --- | --- | --- |
| `try { ... }` alone | passes | passes |
| `try { ... } catch (err) {}` | **fails** | passes |
| `try { ... } // done` | **fails** | passes |
| not wrapped at all (seed) | fails | fails |
| only one statement wrapped | fails | fails |

The `catch` case matters most: the note above tells campers to ignore the red syntax error, so adding a `catch` to silence it is the likeliest thing they will try, and they would get a failure message that does not explain why. Dropping the anchor still pins both statements inside the `try`.

Second, `code.replace(/\s+/g, '')` is what `__helpers.removeWhiteSpace` does. The extra `;?` after `trim())` makes the check tolerant of a dropped semicolon, since the camper is retyping these two lines.

```suggestion
assert.match(__helpers.removeWhiteSpace(code), /try\{constresult=findBook\(bookInput\.value\.trim\(\)\);?statusDisplay\.textContent=result;?\}/);
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Library Book Search</title>
    <link rel="stylesheet" href="styles.css" />
  </head>
  <body>
    <div>
      <h1>Find book title</h1>
      <input type="text" id="bookTitleInput" placeholder="Enter book title">
      <button id="search">Search</button>
      <p id="statusMessage"></p>
      <p id="scanStatus"></p>
    </div>
    <script src="script.js"></script>
  </body>
</html>
```

```css
body {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
}

input {
  padding: 0.5rem;
  width: 200px;
}

button {
  padding: 0.5rem 1rem;
  cursor: pointer;
  background-color: #000000;
  color: white;
}

#statusMessage {
  margin-top: 1.5rem;
  font-weight: bold;
  height: 1.5rem;
}
```

```js
const library = [
  "The Hobbit", "1984", "Dune", "Fahrenheit 451", "To Kill a Mockingbird",
  "The Great Gatsby", "Moby Dick", "Pride and Prejudice", "Frankenstein"
];

const searchBtn = document.getElementById('search');
const statusDisplay = document.getElementById('statusMessage');

> const bookInput = document.getElementById('bookTitleInput');

Reviewer: ```suggestion
const searchBtn = document.getElementById("search");
const statusDisplay = document.getElementById("statusMessage");
const bookInput = document.getElementById("bookTitleInput");
```

const findBook = (title) => {
  if (title === "") {
    throw new Error("Search field cannot be empty");
  }

  if (!library.includes(title)) {
    throw new Error(title + " not found!");
  }

  return title + " is available to borrow!";
}

> searchBtn.addEventListener('click', () => {

Reviewer: ```suggestion
};

searchBtn.addEventListener("click", () => {
```
